Adding JSON output format

DmitryMK opened the pull request:
Merging into the development branch now, as we previously discussed that after 1.0.0 we will stop merging into main directly.

This PR adds 2 options to address https://github.com/cdisc-org/cdisc-rules-engine/issues/66 :
--output-format
--raw-report

By default it will save JSON with the same structure/fields as the Excel report. But for development purposes it might be needed to get results as they are reported by the engine, the --raw-report flag can be used for this purpose.

I created a GenericReport class from the ExcelReport, which is not inherited by both ExcelReport and JsonReport classes. The main reason is to be consistent, so that in case we change how the report looks like in Excel, it will be consistently updated for the JSON.

DmitryMK commented on cdisc_rules_engine/utilities/excel_report.py:
```diff
@@ -162,3 +51,21 @@ def get_excel_export(self, define_version, cdiscCt, standard, version) -> Workbo
+            logger.error(e)
+            raise e
+        finally:
+            self._template.close()
```
It might be a bad design, but it is opened in ReportFactory (line 41 - self._data_service.read_data(..)). I kept it this way for backward compatibility with tests, but I can rewrite ti, so that the template is opened within a class instance and not passed as a parameter.

DmitryMK commented on core.py:
```diff
@@ -90,6 +73,21 @@ def cli():
+    "-of",
+    "--output-format",
+    default="XLSX",
+    type=click.Choice(["JSON", "XLSX"], case_sensitive=False),
```
Fixed

DmitryMK commented on cdisc_rules_engine/utilities/report_factory.py:
```diff
@@ -0,0 +1,57 @@
+                self._args,
+                self._template,
+            )
+        elif output_type == "JSON":
```
Fixed

DmitryMK commented on cdisc_rules_engine/utilities/report_factory.py:
```diff
@@ -0,0 +1,57 @@
+        output_type = self._args.output_format.upper()
+        if output_type == "XLSX":
+            if self._data_service != None:
+                self._template = self._data_service.read_data(
```
Fixed

DmitryMK commented on cdisc_rules_engine/utilities/base_report.py:
```diff
@@ -0,0 +1,167 @@
+)
+
+
+class BaseReport:
```
Updated, also used abstractmethod decorators for the methods which need to be implemented.

DmitryMK commented on cdisc_rules_engine/utilities/excel_report.py:
```diff
@@ -20,128 +22,15 @@ def __init__(
         elapsed_time: float,
-        report_template: TextIO,
+        args: Validation_args,
+        _template: TextIO,
```
Fixed

DmitryMK commented on core.py:
```diff
@@ -121,6 +121,13 @@ def validate(
+
+    # Validate conditional options
+    logger = logging.getLogger("validator")
+    if raw_report == True and output_format.lower() != "json":
```
Fixed

alexfurmenkov commented on cdisc_rules_engine/utilities/base_report.py:
```diff
@@ -0,0 +1,167 @@
+
+    @abstractmethod
+    def get_export() -> Union[dict, Workbook]:
+        raise NotImplementedError
```
`abstractmethod` decorator and raising a `NotImplementedError` are 2 different ways to do the same thing but they should not be combined. 

The difference is that raising a `NotImplementedError` is the old style whereas `abstractmethod` decorator is the new style. A method marked with `abstractmethod` decorator must not contain any implementation (raising an error is also an implementation), the method can only contain `pass` keyword or a docstring. 

In terms of our project, we tend to use the new style but some examples of raising a `NotImplementedError` still exist in the project and my suggestion would be to unify the style later, not in this PR.

DmitryMK replied:
Aleksei for the explanation, will update it. I think that flake8 will catch it as well, so if we have it, that might help to avoid such issues.

DmitryMK replied:
Updated

alexfurmenkov commented on cdisc_rules_engine/utilities/report_factory.py:
```diff
@@ -0,0 +1,58 @@
+    def get_report_service(self):
+        output_type = self._args.output_format.upper()
+        if output_type == ReportTypes.XLSX.value:
+            if self._data_service != None:
```
self._data_service can't be None since it is a mandatory argument in the class init. We can remove this condition safely.

DmitryMK replied:
It is done for backward compatibility with tests, which provide None as a data_service value. I understand, that probably it will be better to rewrite test, rather than having this condition.

alexfurmenkov replied:
Yes, or in test you just can pass a mock

DmitryMK replied:
Updated the test and remove the check for None from the report_factory.

alexfurmenkov commented on cdisc_rules_engine/utilities/base_report.py:
```diff
@@ -0,0 +1,167 @@
+                "status": "SUCCESS"
+                if validation_result.execution_status == "success"
+                else "SKIPPED",
+            }
```
`Please` use ExecutionStatus enum on lines 148, 9

DmitryMK replied:
Updated it for lines 147, 148, 149.

DmitryMK commented on cdisc_rules_engine/utilities/json_report.py:
```diff
@@ -0,0 +1,41 @@
+from cdisc_rules_engine.utilities.generic_report import GenericReport
+
+
+class JsonReport(GenericReport):
```
Added test_json_export.py

alexfurmenkov reviewed (approved):
Approving. Please change the destination branch to main since we have agreed to merge all open PRs and then start creating PRs to development.

nhaydel reviewed (approved):
Sorry for the late response, didn't realize I requested changes on this. Can we merge this into development though

alexfurmenkov commented:
While the PR was hanging, we have again agreed to merge into development. Can you please change the destination branch and do a git pull from development into this branch?

DmitryMK commented:
@alexfurmenkov will merge to development. I have a problem with xmlschema logging - it prints a lot of messages and I think this was introduced in this branch. So I will merge once I fix this issue.

DmitryMK commented:
@alexfurmenkov @nhaydel I have addressed flake8 issues for the files which were updated in this branch (there are other files which are failing flake8 checks). Mainly I just want to confirm that it is fine that I moved `logging.getLogger(XXX).disabled = True` to logging_service_factory.py.
The problem was that flake8 issued an error that import must be at top level because these statements were between imports in run_validation.py. So I had to move it to a new location.

alexfurmenkov commented:
Got it
